modules/core/shared/src/main/scala/lucuma/core/model/TwilightBoundedNight.scala

```diff
@@ -0,0 +1,232 @@
+    * from the given local datetime for for the given site.
+    *
+    * Throws an exception if there's no sunset or sunrise for the specified
+    * [[lucuma.core.enum.TwilightType]], [[lucuma.core.enum.Site]] and
+    * [[java.time.LocalDateTime]].
+    *
+    * @group Constructors
+    */
+  def fromTwilightTypeAndSiteAndLocalDateTimeUnsafe(
+    twilightType: TwilightType,
+    s:            Site,
+    d:            LocalDateTime
+  ): TwilightBoundedNight =
+    fromTwilightTypeAndSiteAndLocalDateTime(twilightType, s, d).get
+
+  /** Constructs a [[TwilightBoundedNight]] for the observing night
+    * that includes the given time at the specified site.
+    *
+    * Returns None if there's no sunset or sunrise for the specified
+    * [[lucuma.core.enum.TwilightType]] and the observing night
+    * that includes the given time at the specified site.
+    *
+    * @group Constructors
+    */
+  def fromTwilightTypeAndSiteAndInstant(
+    twilightType: TwilightType,
+    s:            Site,
+    i:            Instant
+  ): Option[TwilightBoundedNight] =
+    TwilightBoundedNight.fromTwilightTypeAndObservingNight(twilightType,
+                                                           ObservingNight.fromSiteAndInstant(s, i)
+    )
+
+  /** Constructs a [[TwilightBoundedNight]] for the observing night
+    * that includes the given time at the specified site.
+    *
+    * Throws an exception if there's no sunset or sunrise for the specified
+    * [[lucuma.core.enum.TwilightType]] and the observing night
+    * that includes the given time at the specified site.
+    *
+    * @group Constructors
+    */
+  def fromTwilightTypeAndSiteAndInstantUnsafe(
+    twilightType: TwilightType,
+    s:            Site,
+    i:            Instant
+  ): TwilightBoundedNight =
+    fromTwilightTypeAndSiteAndInstant(twilightType, s, i).get
+
+  /** Returns a program in M that computes a [[TwilightBoundedNight]]
+    * corresponding to the observing night for the instant it is executed.
+    *
+    * @group Constructors
+    */
+  def current[M[_]: Sync](twilightType: TwilightType, s: Site): M[Option[TwilightBoundedNight]] =
+    ObservingNight.current(s).map(_.twilightBounded(twilightType))
+
+  /** @group Typeclass Instances. */
+  implicit val ShowTwilightBoundedNight: Show[TwilightBoundedNight] =
+    Show.fromToString
+
+  /** TwilightBoundedNight is ordered by observing night and bound type.
+    *
+    * @group Typeclass Instances
+    */
+  implicit val OrderTwilightBoundedNight: Order[TwilightBoundedNight] =
+    Order.by(n => (n.toObservingNight, n.twilightType))
+
+}
+
+trait TwilightBoundedNightOptics {
+
+  /** @group Optics */
+  val twilightType: Lens[TwilightBoundedNight, TwilightType] =
+    GenLens[TwilightBoundedNight](_.twilightType)
+
+  /** @group Optics */
+  val observingNight: Lens[TwilightBoundedNight, ObservingNight] =
+    GenLens[TwilightBoundedNight](_.toObservingNight)
```

Comment:
I see this is in the old code and maybe I'm confused, but is this safe?  The constructor is marked `private` but I assume a set will make a `copy` swapping out the `ObservingNight` in the `TwilightBoundedNight` (i.e., they become inconsistent).

Reply:
Thank you for catching this. This is code is not that old, I introduced it a while ago when I migrated this class to `gsp-core`.

I'm tempted to move the `Interval` to a lazy val in the class and make it an `Option`, but then we cannot inherit from `Night` anymore. I'll try fixing the `Lens`es so that they recompute.